File: app/services/job_manager.py
import uuid
from datetime import datetime
from typing import Dict, Optional
from ..models import JobStatus


# In-memory job storage
jobs: Dict[str, dict] = {}


def create_job(request) -> str:
    """Create a new job and return its ID."""
    job_id = str(uuid.uuid4())
    jobs[job_id] = {
        "id": job_id,
        "status": JobStatus.PENDING,
        "request": request.model_dump(),
        "created_at": datetime.utcnow().isoformat(),
        "completed_at": None,
        "result_file": None,
        "error": None,
        "progress": 0,
    }
    return job_id


def update_job(job_id: str, **kwargs):
    """Update job status and metadata."""
    if job_id in jobs:
        jobs[job_id].update(kwargs)


def get_job(job_id: str) -> Optional[dict]:
    """Retrieve job by ID."""
    return jobs.get(job_id)


def list_jobs() -> list:
    """List all jobs."""
    return list(jobs.values())

File: server/src/services/jobManager.js
import { v4 as uuidv4 } from 'uuid';

// In-memory job storage (use Redis for production)
const jobs = new Map();

// Callback for WebSocket notifications
let notifyCallback = null;

export const JobStatus = {
  PENDING: 'pending',
  PROCESSING: 'processing',
  COMPLETED: 'completed',
  FAILED: 'failed',
};

/**
 * Create a new job.
 * @param {Object} request - The poster request
 * @returns {string} The job ID
 */
export function createJob(request) {
  const jobId = uuidv4();
  const job = {
    id: jobId,
    status: JobStatus.PENDING,
    progress: 0,
    message: null,
    error: null,
    request,
    createdAt: new Date().toISOString(),
  };

  jobs.set(jobId, job);
  return jobId;
}

/**
 * Get a job by ID.
 * @param {string} jobId - The job ID
 * @returns {Object|null} The job or null if not found
 */
export function getJob(jobId) {
  return jobs.get(jobId) || null;
}

/**
 * Update a job's status.
 * @param {string} jobId - The job ID
 * @param {Object} updates - The updates to apply
 */
export function updateJob(jobId, updates) {
  const job = jobs.get(jobId);
  if (!job) return;

  Object.assign(job, updates);

  // Notify via WebSocket if callback is set
  if (notifyCallback) {
    notifyCallback(jobId, {
      job_id: jobId,
      status: job.status,
      progress: job.progress,
      message: job.message,
      error: job.error,
      download_url: job.status === JobStatus.COMPLETED ? `/api/posters/${jobId}` : null,
    });
  }
}

/**
 * Set the notification callback for WebSocket updates.
 * @param {Function} callback - The callback function
 */
export function setNotifyCallback(callback) {
  notifyCallback = callback;
}

/**
 * Get all jobs (for debugging).
 * @returns {Array} All jobs
 */
export function getAllJobs() {
  return Array.from(jobs.values());
}

/**
 * Delete a job.
 * @param {string} jobId - The job ID
 */
export function deleteJob(jobId) {
  jobs.delete(jobId);
}
